vllm/sequence.py

```diff
@@ -748,6 +748,34 @@ def __eq__(self, other: object) -> bool:
         return self.embeddings == other.embeddings
 
 
+@dataclass
+class IntermediateTensors:
```

Comment:
one comment. Can we avoid sending the residual?  can we implement the model in a way that the residual does not have to go through the next work? Reasons are that sending residual would trigger an additional send/recv

Reply:
Do you mean not having to send the residual at all? Or coalescing the sends by putting the residuals and hidden states into 1 tensor so only 1 send happens? If you mean the first one, I'm not aware of an easy way to avoid sending it in any case as it's not easily recoverable otherwise. We can do the second though, which falls into the category of optimizing the sends/recvs. Currently, I am just trying to get this PR in ASAP as it's fully functional and tested right now, after which we can prioritize these kinds of performance changes after profiling the bottlenecks. These will be more incremental at that point and easier to merge.